# 230629_ML_01


### iris 데이터 셋 분석 by DecisionTree

In [1]:
from sklearn.datasets import load_iris
import numpy as np

iris = load_iris()
type(iris) # Bunch, 딕셔너리 구현체 -> key-value 형태로 구성

sklearn.utils._bunch.Bunch

In [2]:
iris.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [3]:
X = iris['data']
print(type(X), "// shape : ",X.shape)
# shape -> (150,4) -> (데이터 개수, 데이터포인트의 구성) -> (row수, column 수)

<class 'numpy.ndarray'> // shape :  (150, 4)


In [4]:
X[:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

In [5]:
iris["feature_names"] # 각 feature의 의미

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [6]:
y = iris['target']
print(type(y), y.shape)
y[:5]

<class 'numpy.ndarray'> (150,)


array([0, 0, 0, 0, 0])

In [7]:
np.unique(y, return_counts=True)

(array([0, 1, 2]), array([50, 50, 50], dtype=int64))

In [8]:
iris['target_names'] # index : class, value : class name

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [9]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [10]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [11]:
import pandas as pd

In [12]:
# 데이터 셋을 pd.DataFrame 으로 구성

df = pd.DataFrame(iris['data'], columns=iris['feature_names'])
df.head()

# 이게 위 sklearn.datasets의 X 값.

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [13]:
df['target'] = iris['target']
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [14]:
df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
145       2  
146       2  
147       2  
148       2  
149       2

In [15]:
df['class_name'] = df['target'].apply(lambda idx : iris['target_names'][idx])
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target class_name  
0       0     setosa  
1       0     setosa  
2       0     setosa  
3       0     setosa  
4       0     setosa

In [16]:
df['target'].apply(lambda idx : iris['target_names'][idx])

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: target, Length: 150, dtype: object

In [20]:
# 위와 동일하게
def func(idx):
    return iris['target_names'][idx]
func(0)

'setosa'

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int32  
 5   class_name         150 non-null    object 
dtypes: float64(4), int32(1), object(1)
memory usage: 6.6+ KB


In [22]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [23]:
df.groupby('class_name').mean(numeric_only=True)

sepal length (cm)  sepal width (cm)  petal length (cm)  \
class_name                                                           
setosa                  5.006             3.428              1.462   
versicolor              5.936             2.770              4.260   
virginica               6.588             2.974              5.552   

            petal width (cm)  target  
class_name                            
setosa                 0.246     0.0  
versicolor             1.326     1.0  
virginica              2.026     2.0

In [24]:
from sklearn.tree import DecisionTreeClassifier
# 모델 객체 생성
clf = DecisionTreeClassifier()

In [25]:
clf.fit(X, y) # fit(feature, target)

DecisionTreeClassifier()

In [26]:
# 1개 꽃을 추론
new_data = np.array([[5, 3.5, 1.4, 0.25]]) # 2차원으로 추론 데이터 기입
                                           # X의 shape(150,4), 2차원으로 이루어져 있기 때문에
                                           # 추론 데이터도 2차원으로 넣어줘야 한다.
# 3개 꽃들을 추론
new_data = np.array([[5, 3.5, 1.4, 0.25], 
                     [3, 1.5, 1.4, 0.25], 
                     [7, 3.5, 5.4, 4.25]])
new_data.shape

(3, 4)

In [27]:
# 학습한 모델.predict(추론할 feature) -> label(y, target) 값 반환
pred = clf.predict(new_data)
pred

array([0, 0, 2])

In [28]:
# 결과 후처리 : class->class name, 사용자에게 보여주는 결과값으로 처리.
iris['target_names'][pred]


array(['setosa', 'setosa', 'virginica'], dtype='<U10')

## 학습한 모델이 실제 정답, y 값이 맞는지 확인하는 과정


In [29]:
p = clf.predict(X)
p

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [30]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [31]:
(p==y).sum() # True :1 , False :0
            # 150 == All True
    
# 그러나 X는 이전 예측에서 학습했던 feature 값이기 때문에
# 다시 X로 학습해서 실제 값과 비교하는 것은 의미없다. 
# 다른 Feature 값으로 모델 학습한 비교가 없기 때문에.

150

## scikit-learn의  train_test_split() 함수를 이용해 iris 데이터셋 분할
-  train_test_split() : 하나의 데이터셋을 두개의 세트로 분할 하는 함수

#### 데이터셋 분할시 주의
- 분류 문제의 경우 각 클래스(분류대상)가 **같은 비율**로 나뉘어야 한다. 

In [32]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split # train셋과 test셋으로 분리
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

import numpy as np

In [33]:
iris = load_iris() # 데이터 불러오기

X, y = iris.data, iris.target # data -> Feature, target -> label => Bunch :  .표기법, indexer[]
                                                                  # iris['data'] 이런식으로도 부를 수 있다.
X.shape, y.shape

((150, 4), (150,))

In [34]:
# tranin, test 변수들은 튜플로 받는다. 튜플 언패킹
X_train, X_test, y_train, y_test = train_test_split(X, # features 
                                                    y, # label(target)
                                                    test_size=0.2, # test set의 비율 (default : 0.25)
                                                    stratify=y, # 원본의 class별 비율과 동일한 비율로 나누어지도록 한다.
                                                    shuffle=True, # 나누기 전 랜덤하게 섞는다.
                                                    random_state=0 # random seed 값 설정.
                                                   )

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((120, 4), (30, 4), (120,), (30,))

### Random seed (random_state in train_test_split) :

- numpy.random 모듈
- random 값을 출력할 때 무작위로 값이 출력된다.
- random seed 값을 정하면, 무작위로 나오는 값의 순서가 일정해진다.
- seed 값 : 무작위로 나오는 값의 순서를 일정하게 해주기 위해 설정.

#### train_test_split에서 random_state(random seed) 설정의 의미 :

- 전체 데이터에서 train과 test 데이터로 무작위로 나누고 여러 모델에 학습하여 테스트할 때
- 무작위로 데이터를 뽑아서 각기 다른 데이터 셋으로 모델들을 테스트하는 것은 각각의 모델들의 신뢰성과 연관이 없어지기 때문에
- 무작위로 뽑는 데이터 셋들을 random_state를 통해 모두가 일정하게 무작위로 뽑힌 데이터 셋을 사용하게 함으로써 예측 모델들의 평가의 신뢰성을 주기 위함이다.

In [35]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [36]:
y_train

array([0, 0, 0, 0, 1, 0, 2, 2, 1, 2, 2, 1, 0, 1, 2, 2, 0, 1, 1, 0, 2, 0,
       0, 2, 2, 1, 1, 0, 2, 2, 1, 1, 0, 2, 2, 1, 2, 1, 2, 1, 1, 1, 0, 0,
       1, 1, 2, 2, 1, 0, 2, 2, 0, 0, 1, 1, 0, 0, 1, 2, 0, 0, 1, 1, 2, 1,
       2, 0, 0, 2, 1, 1, 0, 0, 2, 1, 2, 0, 1, 2, 2, 1, 2, 0, 1, 0, 0, 2,
       2, 1, 2, 0, 0, 0, 0, 0, 1, 1, 1, 2, 0, 2, 0, 2, 0, 1, 1, 1, 1, 0,
       2, 2, 0, 1, 1, 2, 0, 2, 2, 2])

In [37]:
np.unique(y, return_counts=True)

(array([0, 1, 2]), array([50, 50, 50], dtype=int64))

In [39]:
np.unique(y_train, return_counts=True)
# stratify=y 이를 통해서 y_train의 class 별 값의 비율이 원본 y의 class 별 값의 비율과 동일하게 나뉘어지도록 한 것.

(array([0, 1, 2]), array([40, 40, 40], dtype=int64))

In [40]:
clf = DecisionTreeClassifier()

In [41]:
clf.fit(X_train, y_train)

DecisionTreeClassifier()

In [42]:
### 평가
# 1. 모델 추정 ==> Test set을 이용 (학습할 때 사용하지않은 데이터셋)
pred = clf.predict(X_test)
pred

array([0, 1, 0, 2, 0, 1, 2, 0, 0, 1, 2, 1, 1, 2, 1, 2, 2, 1, 1, 0, 0, 2,
       2, 1, 0, 1, 1, 2, 0, 0])

In [43]:
y_test

array([0, 1, 0, 2, 0, 1, 2, 0, 0, 1, 2, 1, 1, 2, 1, 2, 2, 1, 1, 0, 0, 2,
       2, 2, 0, 1, 1, 2, 0, 0])

In [44]:
# 정확도를 계산하는 공식.
np.sum(pred == y_test)/y_test.size

0.9666666666666667

In [45]:
# 정확도를 계산하는 메서드
accuracy_score(y_test, pred) # 파라미터 순서 -> (정답, 모델 추정값)

0.9666666666666667

In [46]:
result = confusion_matrix(y_test, pred) # (정답, 모델 추정값)
result.shape, type(result)

((3, 3), numpy.ndarray)

In [47]:
result

# axis=0의 index: 정답(실제)의 class (행, row 가 정답) 
# axis=1의 index: 예측결과의 class (열, column이 모델 추정값)
# 이러한 결과값과 축의 개념은 sklearn 에서 정해진 것. 다른 라이브러리 or 환경에서는 다를 수도 있다.

array([[10,  0,  0],
       [ 0, 10,  0],
       [ 0,  1,  9]], dtype=int64)

=> 대각의 값들은 모델 추정값과 정답이 일치하는 것들의 개수,
**idx2 인 값을 모델 추정값이 idx1로 잘못 추정하여 틀렸기 때문에**,
정확도가 0.966666667 이 나왔다.